File: scripts/aws/destroy-kubernetes-cluster.go
package main

import (
	"log"
	"regexp"
	"sync"
	"time"

	"github.com/pkg/errors"
	"github.com/sirupsen/logrus"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/awserr"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/cloudformation"
	"github.com/aws/aws-sdk-go/service/ec2"
	"github.com/aws/aws-sdk-go/service/eks"
	"github.com/aws/aws-sdk-go/service/iam"
)

//nolint:gochecknoglobals
var deferError error

func (ac *AWSCluster) checkDeferError(err error) bool {
	if err != nil {
		if aerr, ok := err.(awserr.Error); ok {
			switch aerr.Code() {
			case "NoSuchEntity":
			case "ResourceNotFoundException":
			case "ValidationError":
			case "InvalidParameterValue":
			default:
				log.Printf("Error (%s): %s\n", aerr.Code(), aerr.Message())
				deferError = aerr
				return true
			}
			log.Printf("Warning (%s): %s\n", aerr.Code(), aerr.Message())
		} else {
			log.Printf("Error: %s\n", err.Error())
			deferError = aerr
		}
		return true
	}
	return false
}

func (ac *AWSCluster) deleteEksRole(iamClient *iam.IAM, eksRoleName *string) {
	log.Printf("Deleting EKS service role \"%s\"...\n", *eksRoleName)

	policyArn := "arn:aws:iam::aws:policy/AmazonEKSClusterPolicy"
	_, err := iamClient.DetachRolePolicy(&iam.DetachRolePolicyInput{
		RoleName:  eksRoleName,
		PolicyArn: &policyArn,
	})
	ac.checkDeferError(err)

	policyArn = "arn:aws:iam::aws:policy/AmazonEKSServicePolicy"
	_, err = iamClient.DetachRolePolicy(&iam.DetachRolePolicyInput{
		RoleName:  eksRoleName,
		PolicyArn: &policyArn,
	})
	ac.checkDeferError(err)

	_, err = iamClient.DeleteRole(&iam.DeleteRoleInput{
		RoleName: eksRoleName,
	})
	if ac.checkDeferError(err) {
		return
	}

	log.Printf("Role \"%s\" successfully deleted!\n", *eksRoleName)
}

func (ac *AWSCluster) deleteEC2NetworkInterfaces(ec2Client *ec2.EC2, cfClient *cloudformation.CloudFormation, nodesStackName *string) {
	cfResp, err := cfClient.DescribeStacks(&cloudformation.DescribeStacksInput{
		StackName: nodesStackName,
	})
	if ac.checkDeferError(err) {
		return
	}

	var sequrityGroup *string
	for _, output := range cfResp.Stacks[0].Outputs {
		if *output.OutputKey == "NodeSecurityGroup" {
			sequrityGroup = output.OutputValue
			break
		}
	}

	ec2Resp, err := ec2Client.DescribeNetworkInterfaces(&ec2.DescribeNetworkInterfacesInput{
		Filters: []*ec2.Filter{
			{
				Name:   aws.String("group-id"),
				Values: []*string{sequrityGroup},
			},
		},
	})

	if ac.checkDeferError(err) {
		return
	}

	for _, networkInterface := range ec2Resp.NetworkInterfaces {
		_, err := ec2Client.DeleteNetworkInterface(&ec2.DeleteNetworkInterfaceInput{
			NetworkInterfaceId: networkInterface.NetworkInterfaceId,
		})

		ac.checkDeferError(err)
	}
}

//nolint:funlen
func (ac *AWSCluster) deleteEksClusterVpc(cfClient *cloudformation.CloudFormation, ec2Client *ec2.EC2, clusterStackName *string) {
	log.Printf("Deleting Amazon EKS Cluster VPC \"%s\"...\n", *clusterStackName)

	resp, err := cfClient.DescribeStacks(&cloudformation.DescribeStacksInput{
		StackName: clusterStackName,
	})
	if ac.checkDeferError(err) {
		return
	}
	outputsMap := outputsToMap(resp.Stacks[0].Outputs)

	output, err := ec2Client.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{
		Filters: []*ec2.Filter{
			{
				Name:   aws.String("vpc-id"),
				Values: []*string{outputsMap.VpcId},
			},
		},
	})
	if !ac.checkDeferError(err) {
		for _, sg := range output.SecurityGroups {
			if *sg.GroupName != "default" {
				log.Printf("Deleting EC2 Security Group %s...", *sg.GroupId)
				_, err = ec2Client.DeleteSecurityGroup(&ec2.DeleteSecurityGroupInput{GroupId: sg.GroupId})
				ac.checkDeferError(err)
			}
		}
	}

	_, err = ec2Client.DeleteVpc(&ec2.DeleteVpcInput{VpcId: outputsMap.VpcId})
	ac.checkDeferError(err)

	stackId := resp.Stacks[0].StackId

	_, err = cfClient.DeleteStack(&cloudformation.DeleteStackInput{
		StackName: clusterStackName,
	})
	if ac.checkDeferError(err) {
		return
	}

	for {
		resp, err := cfClient.DescribeStacks(&cloudformation.DescribeStacksInput{
			StackName: stackId,
		})
		if ac.checkDeferError(err) {
			if err.(awserr.Error).Code() == "ValidationError" {
				log.Printf("Cluster VPC \"%s\" successfully deleted!\n", *clusterStackName)
			}
			return
		}

		switch *resp.Stacks[0].StackStatus {
		case "DELETE_COMPLETE":
			log.Printf("Cluster VPC \"%s\" successfully deleted!\n", *clusterStackName)
			return
		case "DELETE_IN_PROGRESS":
			time.Sleep(requestInterval)
		default:
			log.Printf("Error: Unexpected stack status: %s\n", *resp.Stacks[0].StackStatus)
			deferError = errors.Errorf("unexpected stack status: %s", *resp.Stacks[0].StackStatus)
			return
		}
	}
}

func (ac *AWSCluster) deleteEksCluster(eksClient *eks.EKS, clusterName *string) {
	log.Printf("Deleting Amazon EKS Cluster \"%s\"...\n", *clusterName)

	_, err := eksClient.DeleteCluster(&eks.DeleteClusterInput{
		Name: clusterName,
	})
	if ac.checkDeferError(err) {
		return
	}

	for {
		resp, err := eksClient.DescribeCluster(&eks.DescribeClusterInput{
			Name: clusterName,
		})

		if err != nil && err.(awserr.Error).Code() == "ResourceNotFoundException" {
			log.Printf("EKS Cluster \"%s\" successfully Deleted!\n", *clusterName)
			return
		}

		if ac.checkDeferError(err) {
			return
		}

		switch *resp.Cluster.Status {
		case "DELETING":
			time.Sleep(requestInterval)
		default:
			log.Printf("Error: Unexpected cluster status: %s\n", *resp.Cluster.Status)
			deferError = errors.Errorf("unexpected cluster status: %s", *resp.Cluster.Status)
			return
		}
	}
}

func (ac *AWSCluster) deleteEksEc2KeyPair(ec2Client *ec2.EC2, keyPairName *string) {
	log.Printf("Deleting Amazon EC2 key pair \"%s\"...\n", *keyPairName)
	_, err := ec2Client.DeleteKeyPair(&ec2.DeleteKeyPairInput{
		KeyName: keyPairName,
	})
	if ac.checkDeferError(err) {
		return
	}

	log.Printf("Amazon EC2 key pair \"%s\" successfully Deleted!\n", *keyPairName)
}

func (ac *AWSCluster) deleteEksWorkerNodes(cfClient *cloudformation.CloudFormation, nodesStackName *string) {
	log.Printf("Deleting Amazon EKS Worker Nodes...\n")

	resp, err := cfClient.DescribeStacks(&cloudformation.DescribeStacksInput{
		StackName: nodesStackName,
	})
	if ac.checkDeferError(err) {
		return
	}

	stackId := resp.Stacks[0].StackId

	_, err = cfClient.DeleteStack(&cloudformation.DeleteStackInput{
		StackName: nodesStackName,
	})
	if ac.checkDeferError(err) {
		return
	}

	for {
		resp, err := cfClient.DescribeStacks(&cloudformation.DescribeStacksInput{
			StackName: stackId,
		})
		if ac.checkDeferError(err) {
			return
		}

		switch *resp.Stacks[0].StackStatus {
		case "DELETE_COMPLETE":
			log.Printf("EKS Worker Nodes \"%s\" successfully deleted!\n", *nodesStackName)
			return
		case "DELETE_IN_PROGRESS":
			time.Sleep(requestInterval)
		default:
			log.Printf("Error: Unexpected stack status: %s\n", *resp.Stacks[0].StackStatus)
			deferError = errors.Errorf("unexpected stack status: %s", *resp.Stacks[0].StackStatus)
			return
		}
	}
}

// DeleteAWSKubernetesCluster - deleting AWS cluster instances
func (ac *AWSCluster) DeleteAWSKubernetesCluster() error {
	sess := session.Must(session.NewSession())
	iamClient := iam.New(sess)
	eksClient := eks.New(sess)
	cfClient := cloudformation.New(sess)
	ec2Client := ec2.New(sess)

	// Deleting Amazon EKS Cluster
	clusterName := awsClusterPrefix + ac.serviceSuffix
	ac.deleteEksCluster(eksClient, &clusterName)

	nodesStackName := awsNodesStackPrefix + ac.serviceSuffix
	// Deleting EC2 Network Interfaces to allow instance to be properly released
	ac.deleteEC2NetworkInterfaces(ec2Client, cfClient, &nodesStackName)

	// Deleting Amazon EKS Worker Nodes
	ac.deleteEksWorkerNodes(cfClient, &nodesStackName)

	// Deleting Amazon EKS Cluster VPC
	clusterStackName := awsClusterStackPrefix + ac.serviceSuffix
	ac.deleteEksClusterVpc(cfClient, ec2Client, &clusterStackName)

	// Deleting Amazon Roles and Keys
	eksRoleName := awsRolePrefix + ac.serviceSuffix
	ac.deleteEksRole(iamClient, &eksRoleName)
	keyPairName := awsKeyPairPrefix + ac.serviceSuffix
	ac.deleteEksEc2KeyPair(ec2Client, &keyPairName)

	return deferError
}

// DeleteAllKubernetesClusters - removes all aws clusters older than saveDuration and filtered by name
func DeleteAllKubernetesClusters(saveDuration time.Duration, namePattern string) {
	sess := session.Must(session.NewSession())
	cfClient := cloudformation.New(sess)

	var filter = []*string{aws.String("CREATE_IN_PROGRESS"), aws.String("CREATE_FAILED"), aws.String("CREATE_COMPLETE"), aws.String("ROLLBACK_IN_PROGRESS"), aws.String("ROLLBACK_FAILED"), aws.String("ROLLBACK_COMPLETE"), aws.String("DELETE_IN_PROGRESS"), aws.String("DELETE_FAILED"), aws.String("UPDATE_IN_PROGRESS"), aws.String("UPDATE_COMPLETE_CLEANUP_IN_PROGRESS"), aws.String("UPDATE_COMPLETE"), aws.String("UPDATE_ROLLBACK_IN_PROGRESS"), aws.String("UPDATE_ROLLBACK_FAILED"), aws.String("UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS"), aws.String("UPDATE_ROLLBACK_COMPLETE"), aws.String("REVIEW_IN_PROGRESS")}
	stacks, err := cfClient.ListStacks(&cloudformation.ListStacksInput{StackStatusFilter: filter})
	if err != nil {
		logrus.Infof("AWS EKS Error: %v", err)
		return
	}

	var wg sync.WaitGroup
	i := 0
	for _, stack := range stacks.StackSummaries {
		stackName := aws.StringValue(stack.StackName)
		if stackName[:len(awsClusterStackPrefix)] != "nsm-srv" {
			continue
		}

		if aws.StringValue(stack.StackStatus) == "DELETE_COMPLETE" {
			continue
		}

		if matched, err := regexp.MatchString(namePattern, stackName[len(awsClusterStackPrefix):]); err != nil || !matched {
			logrus.Infof("Skip cluster: %s (matched: %v; err: %v)", stackName[len(awsClusterStackPrefix):], matched, err)
			continue
		}

		if stack.CreationTime.After(time.Now().Add(-1 * saveDuration)) {
			logrus.Infof("Skip cluster: %s (created: %v)", stackName[len(awsClusterStackPrefix):], stack.CreationTime)
			continue
		}

		wg.Add(1)

		go func() {
			defer wg.Done()
			for att := 0; att < 3; att++ {
				logrus.Infof("Deleting %s (created %v), attempt %d", stackName[len(awsClusterStackPrefix):], stack.CreationTime, att+1)
				err := NewAWSCluster(stackName[len(awsClusterStackPrefix):]).DeleteAWSKubernetesCluster()
				if err == nil {
					break
				}
			}
		}()

		i++
		if i%5 == 0 {
			wg.Wait() // Guard from AWS Throttling error
		}
	}
	wg.Wait()
}

File: scripts/aws/destroy-kubernetes-cluster.py
#!/usr/bin/env python

import boto3
import os
import sys
import time

from botocore.exceptions import ClientError
from threading import Thread

eks_client = boto3.client('eks')
cf_client = boto3.client('cloudformation')
iam_client = boto3.client('iam')
ec2_client = boto3.client('ec2')

def delete_eks_role(eks_role_name):
    print("Deleting EKS service role \"%s\"..." % eks_role_name)

    try:
        iam_client.detach_role_policy(
            RoleName = eks_role_name,
            PolicyArn = 'arn:aws:iam::aws:policy/AmazonEKSClusterPolicy'
        )
    except ClientError as e:
        print("Warning: %s" % e.response['Error']['Message'])

    try:
        iam_client.detach_role_policy(
            RoleName = eks_role_name,
            PolicyArn = 'arn:aws:iam::aws:policy/AmazonEKSServicePolicy'
        )
    except ClientError as e:
        print("Warning: %s" % e.response['Error']['Message'])

    try:
        iam_client.delete_role(
            RoleName = eks_role_name
        )
    except ClientError as e:
        print("Error: %s" % e.response['Error']['Message'])
        return

    print("Role \"%s\" successfully deleted!" % (eks_role_name))

def delete_eks_cluster_vpc(cluster_stack_name):

    sys.stdout.write("Deleting Amazon EKS Cluster VPC \"%s\"...\n" % cluster_stack_name)

    try:
        cf_client.delete_stack(
            StackName=cluster_stack_name
        )
    except ClientError as e:
        print("Error: %s" % e.response['Error']['Message'])
        return

    while True:
        try:
            response = cf_client.describe_stacks(StackName=cluster_stack_name)
        except ClientError as e:
            print("Error: %s" % e.response['Error']['Message'])
            return

        if response[u'Stacks'][0][u'StackStatus'] == "DELETE_COMPLETE":
            break
        elif response[u'Stacks'][0][u'StackStatus'] == "DELETE_IN_PROGRESS":
            time.sleep(1)
        else:
            print("Error: Unexpected stack status: %s" % response[u'Stacks'][0][u'StackStatus'])
            return

    print("Cluster VPC \"%s\" successfully deleted!" % cluster_stack_name)

def delete_eks_cluster(cluster_name):
    sys.stdout.write("Deleting Amazon EKS Cluster \"%s\"...\n" % cluster_name)
    try:
        eks_client.delete_cluster(
            name = cluster_name,
        )
    except ClientError as e:
        print("Error: %s" % e.response['Error']['Message'])
        return

    while True:
        try:
            response = eks_client.describe_cluster(name=cluster_name)
        except ClientError as e:
            break

        if response[u'cluster'][u'status'] == "DELETING":
            time.sleep(1)
        else:
            print("Error: Unexpected cluster status: %s" % response[u'cluster'][u'status'])
            return

    print("EKS Cluster \"%s\" successfully deleted!" % cluster_name)

def delete_eks_ec2_key_pair(key_name):
    print("Deleting Amazon EC2 key pair \"%s\"..." % key_name)

    try:
        ec2_client.delete_key_pair(
            KeyName=key_name
        )
    except ClientError as e:
        print("Error: %s" % e.response['Error']['Message'])
        return

    print("Amazon EC2 key pair \"%s\" successfully deleted!" % key_name)

def delete_eks_worker_nodes(nodes_stack_name):
    sys.stdout.write("Deleting Amazon EKS Worker Nodes...\n")

    try:
        cf_client.delete_stack(
            StackName=nodes_stack_name,
        )
    except ClientError as e:
        print("Error: %s" % e.response['Error']['Message'])
        return

    while True:
        try:
            response = cf_client.describe_stacks(StackName=nodes_stack_name)
        except ClientError as e:
            print("Error: %s" % e.response['Error']['Message'])
            return

        if response[u'Stacks'][0][u'StackStatus'] == "DELETE_COMPLETE":
            break
        elif response[u'Stacks'][0][u'StackStatus'] == "DELETE_IN_PROGRESS":
            time.sleep(1)
        else:
            print("Error: Unexpected stack status: %s" % response[u'Stacks'][0][u'StackStatus'])
            return

    print("EKS Worker Nodes \"%s\" successfully deleted!" % nodes_stack_name)

def main():
    suffix = os.environ.get("NSM_AWS_SERVICE_SUFFIX", "")

    #### Deleting Amazon EKS Worker Nodes
    nodes_stack_name = "nsm-nodes" + suffix
    delete_eks_worker_nodes_thread = Thread(
        target = delete_eks_worker_nodes,
        args = (nodes_stack_name,)
    )
    delete_eks_worker_nodes_thread.start()

    #### Deleting Amazon EKS Cluster
    cluster_name = "nsm" + suffix
    delete_eks_cluster_thread = Thread(
        target = delete_eks_cluster,
        args = (cluster_name,)
    )
    delete_eks_cluster_thread.start()

    #### Deleting Amazon EKS Cluster VPC
    cluster_stack_name = 'nsm-srv' + suffix
    delete_eks_cluster_vpc_thread = Thread(
        target = delete_eks_cluster_vpc,
        args = (cluster_stack_name,)
    )
    delete_eks_cluster_vpc_thread.start()

    #### Wait for deletion complete
    delete_eks_worker_nodes_thread.join()
    delete_eks_cluster_thread.join()
    delete_eks_cluster_vpc_thread.join()

    #### Deleting keys and roles
    key_pair_name = "nsm-key-pair" + suffix
    delete_eks_ec2_key_pair(key_pair_name)
    eks_role_name = "nsm-role" + suffix
    delete_eks_role(eks_role_name)

if __name__ == '__main__':
    main()
